### PR TITLE
Type "rs\n" to restart tests, fixes #871

### LINKED ISSUE #871: Add a key-press to re-run the specs under --watch
I love --watch. It's great. But there's a few situations where I need to re-run the tests, and it doesn't catch them (i.e. changing an implementation file, as opposed to a test-file, see #870); it'd be nice if hitting the spacebar with the Terminal focused (or something along those lines) forced a re-run of the tests.

Along the same lines, it might be nice to be able to configure a minimum timeout to re-run them *anyway*, so that the window is always up-to-date, even if it misses a change. (Obviously more tenable for small projects than big ones … but isn't that *already* the case with `--watch`?)

### PR DESCRIPTION
**Pull request by broofa**
### Description of the Change

Add a keyboard shortcut (`rs\n`) for manually restarting tests when running the CLI in watch (`--watch`) mode.Although the `--watch` feature is somewhat controversial (see #1780), hopefully it's okay to make minor improvements while that debate is being hashed out? 

To test this feature:
1. `mocha -w` to run CLI in watch mode
2. Type [r] [s] [Enter] 
3. Notice that tests restart

### Alternate Designs

This feature mimics the behavior of the popular `nodemon` module, so should be familiar to many developers.  (And, in fact, the code is cribbed pretty much straight from nodemon)

### Why should this be in core?

This is a minor tweak to behavior already in core (`--watch`).  Although the watch feature is somewhat controversial (see #1780), hopefully there's no reason not to make minor improvements as long as it's supported.

### Benefits

Provides behavior that many developers are familiar with in similar tool(s) - specifically `nodemon`.

### Possible Drawbacks

This may make using stdin in the CLI (watch mode) problematic in the future, however that seems like a pretty abstract concern, and mocha does not appear to do anything with stdin [at this time](https://github.com/mochajs/mocha/search?q=stdin&type=Code)

### Applicable issues

None that I'm aware of.

This is an enhancement (semver "minor" release)

### CODE REVIEW

**Comment by createthis** [comment]
@broofa I am searching for ways to improve my heavily disk-laden test executions due to multiple import/require statements, so I thought I would give your branch a test run.

This code does not appear to work for me. In a multi-file system where my test requires another file, if I modify the required file and type "rs\n" I see the files re-run, but the required file does not appear to have been reloaded from disk.

**Comment by broofa** [comment]
@createthis Can you put up an [SSCCE](http://sscce.org/)?  I tried to repro your issue with the following (very simple) test, but wasn't able to.  Modifying `b.js` (the required file) causes mocha to reload+rerun, without even having to hit "rs\n".  "rs\n" reruns, of course, but since the files we're already reloaded by mocha, it behaves as-expected:

====
`a.js`:
```
let {hi, bye} = require('./b');

describe('HI', hi);
describe('BYE', bye);
```

`b.js`:
```
exports.hi = () => it('Hello', () => true);
exports.bye = () => it('Good Bye', () => true);
```

Command line:
```
$ mocha --watch a.js
```

Looking at the source code, I think this PR is doing the right thing - it calls `rerun()`, which calls the [purge()](https://github.com/mochajs/mocha/blob/833e7b0512ee5ba16fc3b2c63b5b88404bf37119/lib/cli/watch-run.js#L57) function to have Mocha unload the files in question.

**Comment by createthis** [comment]
@broofa You right. This was user error. I was editing a `jsx` file without `--extension js,jsx`. I expected `rs\n` to reload everything but apparently it only reloads what it's watching.

**Comment by broofa** [comment]
@boneskull @craigtaub Any comment on the likelihood of this PR being accepted?  'Just wondering if I should bother figuring out how to resolve the conflicts (not quite as trivial a task as I though initially.)

**Comment by craigtaub** [comment]
Sorry took a while to response. Nice change @broofa . 

So relatively recently we actually swapped to using `chokidar` for handling watching. If you could find a way to do it with that it would be ideal.

**Comment by boneskull** [comment]
thanks!! yes, this would be a good add if you want to work on it. otherwise, maybe someone will pick up where you left off.

**Comment by broofa** [comment]
Cool.  I'm busy with work for the next month or so, but I'll try to find some time to patch this up.  If someone wants to help out with it that'd be cool, too.

**Comment by broofa** [comment]
@boneskull Can you take another look at this?  I've resolved conflicts and added a short description in "index.md".  (Can't seem to build the site, though because node-gyp isn't installing properly for me at the moment.  Also, I'm not sure why README.md is showing as changed - that seems to be an artifact of merging master. :-/)

File: docs/index.md [hunk 1]
```diff
@@ -1578,17 +1580,17 @@ mocha.setup({
 
 ### Browser-specific Option(s)
 
-Browser Mocha supports many, but not all [cli options](#command-line-usage).  
+Browser Mocha supports many, but not all [cli options](#command-line-usage).
 To use a [cli option](#command-line-usage) that contains a "-", please convert the option to camel-case, (eg. `check-leaks` to `checkLeaks`).
 
 #### Options that differ slightly from [cli options](#command-line-usage):
 
-`reporter` _{string|constructor}_
```
Comment (craigtaub) [inline docs/index.md]: Could you revert these? Removes the new-line so its harder to read.

**Comment by craigtaub** [comment]
Small comment, but really it LGTM

File: docs/index.md (hunk 1 above)
Comment (broofa) [reply docs/index.md]: Done.  Sorry about that.  My editor removes trailing whitespace automatically.  'Didn't realize it was significant in Markdown.

**Comment by broofa** [comment]
@boneskull @craigtaub I think this ready to go.  I'll be out of the country for the next couple of weeks.  It'd be nice to see this merged before more conflicts develop.  Thanks!

**Comment by broofa** [comment]
Ping: _"It'd be nice to see this merged before more conflicts develop. Thanks!"_

**Comment by craigtaub** [comment]
Tested locally. Seems good.

**Pull request merged**